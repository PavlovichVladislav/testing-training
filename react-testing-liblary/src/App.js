import { useEffect, useState } from "react";
import "./App.css";

function App() {
   const [data, setData] = useState(null);
   const [toggle, setToggle] = useState(false);
   const [value, setValue] = useState(false);

   const onClick = () => setToggle((prev) => !prev);

   useEffect(() => {
      setTimeout(() => {
         setData({});
      }, 1000);
   }, []);

   return (
      <div className="App">
        <h1 data-testId="value-elem">{value}</h1>
         {toggle && <div data-testid="toggle-elem">toggle</div>}
         {data && <div style={{ color: "red" }}>data</div>}
         <h1>Hello world</h1>
         <button onClick={onClick} data-testid="toggle-button">
            click me
         </button>
         <input
            type="text"
            placeholder="input value"
            value={value}
            onChange={(e) => setValue(e.target.value)}
         />
      </div>
   );
}

export default App;
