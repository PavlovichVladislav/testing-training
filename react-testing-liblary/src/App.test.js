import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import App from "./App";

describe("Test app", () => {
   test("renders layout`", () => {
      render(<App />);
      const helloWorldElem = screen.getByText(/hello world/i);
      const btn = screen.getByRole("button");
      const input = screen.getByPlaceholderText(/input value/i);

      expect(helloWorldElem).toBeInTheDocument();
      expect(btn).toBeInTheDocument();
      expect(input).toMatchSnapshot();
   });

   test("elem does not exist on page`", () => {
      render(<App />);
      const helloElem = screen.queryByText(/hello2/i);
      expect(helloElem).toBeNull();
   });

   test("the element appears on the page after some time`", async () => {
      render(<App />);
      screen.debug();
      const dataElem = await screen.findByText(/data/i);
      expect(dataElem).toBeInTheDocument();
      expect(dataElem).toHaveStyle({ color: "red" });
      screen.debug();
   });

   test("CLICK EVENT`", async () => {
      render(<App />);
      const btn = screen.getByTestId("toggle-button");

      expect(screen.queryByTestId("toggle-elem")).toBeNull();
      fireEvent.click(btn);
      expect(screen.queryByTestId("toggle-elem")).toBeInTheDocument();
      fireEvent.click(btn);
      expect(screen.queryByTestId("toggle-elem")).toBeNull();
   });

   test("INPUT EVENT`", async () => {
      render(<App />);
      const input = screen.getByPlaceholderText(/input value/i);
      expect(screen.queryByTestId("value-elem")).toContainHTML('');

      // искусственное событие
      fireEvent.input(input, {
         target: {value: '123'}
      });
      
      // пользовательское событие
      userEvent.type(input, '123')

      expect(screen.queryByTestId("value-elem")).toContainHTML('123123');
   });
});
