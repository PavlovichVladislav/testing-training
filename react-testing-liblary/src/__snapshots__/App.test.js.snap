// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test app renders layout\` 1`] = `
<input
  placeholder="input value"
  type="text"
  value="false"
/>
`;
